rust/src/checkpoints.rs

```diff
@@ -67,187 +67,170 @@ pub enum CheckPointWriterError {
     },
 }
 
-impl From<CheckPointWriterError> for ArrowError {
-    fn from(error: CheckPointWriterError) -> Self {
+impl From<CheckpointError> for ArrowError {
+    fn from(error: CheckpointError) -> Self {
         ArrowError::from_external_error(Box::new(error))
     }
 }
 
-/// Struct for writing checkpoints to the delta log.
-pub struct CheckPointWriter {
-    table_uri: String,
-    delta_log_uri: String,
-    last_checkpoint_uri: String,
-    storage: Box<dyn StorageBackend>,
+/// Loads table from given `table_uri` at given `version` and creates checkpoints for it.
+pub async fn create_checkpoint_from_table_uri(
+    table_uri: &str,
+    version: DeltaDataTypeVersion,
+) -> Result<(), CheckpointError> {
+    let table = open_table_with_version(table_uri, version).await?;
+    create_checkpoint(
+        version,
+        table.get_state(),
+        table.storage.as_ref(),
+        table_uri,
+    )
+    .await?;
+    Ok(())
 }
 
-impl CheckPointWriter {
-    /// Creates a new CheckPointWriter.
-    pub fn new(table_uri: &str, storage: Box<dyn StorageBackend>) -> Self {
-        let delta_log_uri = storage.join_path(table_uri, "_delta_log");
-        let last_checkpoint_uri = storage.join_path(delta_log_uri.as_str(), "_last_checkpoint");
-
-        Self {
-            table_uri: table_uri.to_string(),
-            delta_log_uri,
-            last_checkpoint_uri,
-            storage,
-        }
-    }
-
-    /// Creates a new CheckPointWriter for the table URI.
-    pub fn new_for_table_uri(table_uri: &str) -> Result<Self, CheckPointWriterError> {
-        let storage_backend = storage::get_backend_for_uri(table_uri)?;
-
-        Ok(Self::new(table_uri, storage_backend))
+/// Loads given `table` at given `version` and creates checkpoints for it.
+/// If given table's version does not match given `version`, then `table` is re-loaded with appropriate one.
+/// Once checkpoint is created, then the `table` is updated to the latest version.
+pub async fn create_checkpoint_from_table(
+    table: &mut DeltaTable,
+    version: DeltaDataTypeVersion,
+) -> Result<(), CheckpointError> {
+    if table.version != version {
+        table.load_version(version).await?;
     }
+    create_checkpoint(
+        version,
+        table.get_state(),
+        table.storage.as_ref(),
+        &table.table_uri,
+    )
+    .await?;
+    table.update_incremental().await?;
+    Ok(())
+}
```

Comment:
Had a conversation with @xianwill on the behaviour of this call, some users might don't want to update to the latest version. But the `table.update_incremental().await?;` has been added in case the `table.load_version(version).await?;` was a time travel. So in this case the returned table will end up always with `>= pervious_version`

Reply:
Opened an issue https://github.com/delta-io/delta-rs/issues/385

cc @houqp